### PR TITLE
TEST-50-DISSECT: always use systemd-notify to send notify message

### LINKED ISSUE #35619: TEST-50-DISSECT.dissect is unstable
https://github.com/systemd/systemd/actions/runs/12327944793/job/34410554367?pr=35613
```
[   35.183790] TEST-50-DISSECT.sh[950]: + systemd-run --wait -p RootImage=/tmp/TEST-50-IMAGES.mnm/minimal_0.raw -p NotifyAccess=all --service-type=notify --pipe sh -c 'echo READY=1 | ncat --unixsock --udp $NOTIFY_SOCKET --source /run/notify && env'
[   35.194600] TEST-50-DISSECT.sh[951]: + grep NOTIFY_SOCKET=/run/host/notify
[   35.293424] device-mapper: table: 253:0: verity: Root hash verification failed (-ENOKEY)
[   35.299418] device-mapper: ioctl: error adding target to table
[   35.675771] kauditd_printk_skb: 118 callbacks suppressed
[   35.675777] audit: type=1130 audit(1734163644.701:254): pid=1 uid=0 auid=4294967295 ses=4294967295 msg='unit=run-r34eab927d20c481ea4dc10a5c22d33a3 comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=failed'
[   35.682441] audit: type=1338 audit(1734163644.709:255): module=verity op=dtr ppid=2 pid=319 auid=4294967295 uid=0 gid=0 euid=0 suid=0 fsuid=0 egid=0 sgid=0 fsgid=0 tty=(none) ses=4294967295 comm="kworker/u8:4" exe=(null) dev=253:0 error_msg='success' res=1
[   35.691833] TEST-50-DISSECT.sh[951]: NOTIFY_SOCKET=/run/host/notify
[   35.694329] TEST-50-DISSECT.sh[551]: + echo 'Subtest /usr/lib/systemd/tests/testdata/units/TEST-50-DISSECT.dissect.sh failed'
[   35.694947] TEST-50-DISSECT.sh[551]: Subtest /usr/lib/systemd/tests/testdata/units/TEST-50-DISSECT.dissect.sh failed
```
The test case is added by 284dd31e9d7d25e8a0bdfee60cf938ab961f2a7a.

cc @DaanDeMeyer

### PR DESCRIPTION
**Pull request by yuwata**
Follow-up for d0a63cf0412ccb9b4edd1f6048e25aa7f68b0133.
Fixes #35619.

### CODE REVIEW

**Review by DaanDeMeyer — APPROVED** [review]
Having a non flaky test seems more important than testing the notify socket protocol without systemd-notify in this case. We can always add that back in a dedicated test for systemd-notify. This test is about making sure the notify socket is mounted at the right location in the sandbox, not about the notify socket protocol itself. And I double checked that systemd-notify uses the BARRIER=1 protocol which should eliminate all races.

File: test/units/TEST-50-DISSECT.dissect.sh [hunk 1]
```diff
@@ -89,32 +89,40 @@ systemd-run \
...
-            systemd-notify --pid=auto --ready
-            [[ "$$NOTIFY_SOCKET" == "/run/host/notify" ]]
-            [[ "$(env)" =~ "NOTIFY_SOCKET=/run/host/notify" ]]
-            test -S /run/host/notify
-        '
-
-    btrfs subvolume delete /test-dissect-btrfs-snapshot
-fi
+# Same test with RootDirectory=
+mkdir -p /tmp/img
```
Comment (daandemeyer) [inline test/units/TEST-50-DISSECT.dissect.sh]: Seems `unsquashfs` will create the directory itself and this causes the test to fail.

**Comment by bluca** [comment]
> Having a non flaky test seems more important than testing the notify socket protocol without systemd-notify in this case. We can always add that back in a dedicated test for systemd-notify. This test is about making sure the notify socket is mounted at the right location in the sandbox, not about the notify socket protocol itself. And I double checked that systemd-notify uses the BARRIER=1 protocol which should eliminate all races.

If that's all it needs to do, then just do `test -S` instead, that will be enough to verify the socket is available. The builds are already very slow as it is, adding the systemd package and all its dependencies to this image will make them even slower.

**Comment by DaanDeMeyer** [comment]
> > Having a non flaky test seems more important than testing the notify socket protocol without systemd-notify in this case. We can always add that back in a dedicated test for systemd-notify. This test is about making sure the notify socket is mounted at the right location in the sandbox, not about the notify socket protocol itself. And I double checked that systemd-notify uses the BARRIER=1 protocol which should eliminate all races.
> 
> If that's all it needs to do, then just do `test -S` instead, that will be enough to verify the socket is available. The builds are already very slow as it is, adding the systemd package and all its dependencies to this image will make them even slower.

Without this PR: https://github.com/systemd/systemd/actions/runs/12366308331/job/34512776149
With this PR: https://github.com/systemd/systemd/actions/runs/12367485430/job/34515871742?pr=35652

No difference at all in image build time and test runtime

**Comment by bluca** [comment]
An extra 10 seconds here, and almost 20M increase in size

**Comment by yuwata** [comment]
> An extra 10 seconds here, and almost 20M increase in size

That Should not impact anything.
E.g. TEST-73-LOCALE takes about 500 sec on fedora, and 1000 sec on rawhide with sanitizers. So, extra 10 seconds on building image should not impact whole test performance.
Also, the uncompressed main image size is 9.1G for fedora rawhide. Additional 20M can be negligible.

**Comment by bluca** [comment]
> > An extra 10 seconds here, and almost 20M increase in size
> 
> That Should not impact anything. E.g. TEST-73-LOCALE takes about 500 sec on fedora, and 1000 sec on rawhide with sanitizers. So, extra 10 seconds on building image should not impact whole test performance. Also, the uncompressed main image size is 9.1G for fedora rawhide. Additional 20M can be negligible.

Those are test runs times, I am talking about build times, which are paid every time a rebuild is done, given these images are always built. And the minimal image is supposed to be minimal, it's not really minimal anymore if the whole kitchen sink is thrown into it.
Again if the purpose is to check the socket being mounted, just do `test -S`? Simpler and less error prone.

**Comment by yuwata** [comment]
OK. Made the minimal image minimal again. PTAL.

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]
I don't think this will work, the socket won't be at /run/host/notify unless Type=notify is used. And if Type=notify is used, `systemd-run` will fail with a protocol error unless the service properly sends `READY=1`

**Comment by DaanDeMeyer** [comment]
So given we can't properly implement the notify socket protocol race free in bash and we have to use `Type=notify`, maybe we can protocol error exit code from `systemd-run` if it's different from the exit code from `test -S`?

**Comment by bluca** [comment]
> I don't think this will work, the socket won't be at /run/host/notify unless Type=notify is used. And if Type=notify is used, `systemd-run` will fail with a protocol error unless the service properly sends `READY=1`

Mmh should we just mount it unconditionally then? In the end the notify protocol is also useful for sending status updates that are shown by "systemctl status"

**Comment by DaanDeMeyer** [comment]
Actually I misspoke, it seems to be gated purely on the value of `NotifyAccess=`, so this might work after all

**Review by bluca — APPROVED** [review]

**Pull request merged**